### PR TITLE
DDP-6284 autocomplete picklist

### PR DESCRIPTION
**Pull request by aglushko-broad**

### CODE REVIEW

File: ddp-workspace/projects/ddp-sdk/src/lib/pipes/searchHighlight.pipe.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { Pipe, PipeTransform } from '@angular/core';
+
+@Pipe({
+    name: 'searchHighlight'
+})
+export class SearchHighlightPipe implements PipeTransform {
+    transform(text: string, search: string): string {
+        const pattern = search
+            .replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, '\\$&')
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/pipes/searchHighlight.pipe.ts:11]: A comment might be a good idea here.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/autocompleteActivityPicklistQuestion.component.ts [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+    private generateSuggestions(): Suggestion[] {
+        return this.block.picklistOptions.reduce((acc, currentValue) => {
+            if (currentValue.allowDetails) return acc;
+            const suggestion: Suggestion = {
+                label: currentValue.optionLabel,
+                value: currentValue.stableId,
+            };
+            acc.push(suggestion);
+            if (currentValue.nestedOptions) {
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/autocompleteActivityPicklistQuestion.component.ts]: Consider checking explicitly for empty array (here, empty array would evaluate to true), so would be treated as a parent even though no children.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/autocompleteActivityPicklistQuestion.component.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    public ngOnInit(): void {
+        this.allSuggestions = this.generateSuggestions();
+        const customVersionId = this.block.picklistOptions.find(option => option.allowDetails)?.stableId;
+
+        this.inputFormControl.valueChanges.pipe(
+            map(value => typeof value === 'string' ? value.trim() : value),
+            distinctUntilChanged(),
+            debounceTime(200),
+            takeUntil(this.ngUnsubscribe)
+        ).subscribe((value: string | Suggestion) => {
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/autocompleteActivityPicklistQuestion.component.ts]: Do we have to worry about unsubscribing?

Comment (aglushko-broad) [reply ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/autocompleteActivityPicklistQuestion.component.ts]: we do unsubscribe on the line above

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/autocompleteActivityPicklistQuestion.component.ts (hunk 2 above)
Comment (aglushko-broad) [reply ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/autocompleteActivityPicklistQuestion.component.ts]: good idea! done

File: ddp-workspace/projects/ddp-sdk/src/lib/pipes/searchHighlight.pipe.ts (hunk 1 above)
Comment (aglushko-broad) [reply ddp-workspace/projects/ddp-sdk/src/lib/pipes/searchHighlight.pipe.ts:11]: done!

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/autocompleteActivityPicklistQuestion.component.ts [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+                        class="autoCompleteOption"
+                        [value]="suggestion"
+                        [class.parent]="suggestion.isParent"
+                        [class.child]="suggestion.parent">
+                <span [innerHtml]="suggestion.label | searchHighlight: autocompleteInput.value"></span>
+            </mat-option>
+        </mat-autocomplete>
+    </mat-form-field>`,
+    styles: [`
+        .width {
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/autocompleteActivityPicklistQuestion.component.ts]: #minor

Perhaps the class name `full-width` would be more expressive.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/autocompleteActivityPicklistQuestion.component.ts [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    constructor(translate: NGXTranslateService) {
+        super(translate);
+    }
+
+    public ngOnInit(): void {
+        this.allSuggestions = this.generateSuggestions();
+        const customVersionId = this.block.picklistOptions.find(option => option.allowDetails)?.stableId;
+
+        this.inputFormControl.valueChanges.pipe(
+            map(value => typeof value === 'string' ? value.trim() : value),
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/autocompleteActivityPicklistQuestion.component.ts:69]: What kind of type can it be except `string` ?

File: ddp-workspace/projects/ddp-basil/src/assets/i18n/en.json [hunk 6]
```diff
@@ -1,299 +1,300 @@
...
+            "Name": "Name:",
+            "ShortId": "Subject Short ID/HRUID:"
+        },
+        "AdminActionPanel": {
+            "ActivityEdit": "Click the Edit button to temporarily unlock for editing. Your edits will be saved as you go. When you click submit or navigate away, the activity will be locked again automatically.",
+            "EditButton": "Edit"
+        },
+        "Validators": {
+            "DateNavyValidationRule": "Entered date is invalid",
+            "Autocomplete": "No {{control}} is found"
```
Comment (aglushko-broad) [inline ddp-workspace/projects/ddp-basil/src/assets/i18n/en.json:279]: only this line was added

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/activityQuestion.component.spec.ts [hunk 7]
```diff
@@ -1,162 +1,191 @@
...
-  @Output() valueChanged = new EventEmitter();
-  @Input() block;
-  @Input() readonly;
-  @Input() validationRequested;
+    @Output() valueChanged = new EventEmitter();
+    @Input() block;
+    @Input() readonly;
+    @Input() validationRequested;
+    @Input() studyGuid;
+    @Input() activityGuid;
```
Comment (aglushko-broad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/activityQuestion.component.spec.ts:31]: 2 more inputs were added

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/activityQuestion.component.spec.ts [hunk 8]
```diff
@@ -1,162 +1,191 @@
...
-            blockVisibility: []
-          })
-        }));
-    component.ngOnInit();
-    getTestScheduler().flush();
-    fixture.detectChanges();
-    tick(1);
-    expect(getValidationMessageFixture()).toBeFalsy();
-  }));
+    it('expect block validation to appear for validator which returns object as result', fakeAsync(() => {
```
Comment (aglushko-broad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/activityQuestion.component.spec.ts:138]: this new test was added

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/autocompleteActivityPicklistQuestion.component.ts (hunk 4 above)
Comment (aglushko-broad) [reply ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/picklist/autocompleteActivityPicklistQuestion.component.ts]: fixed

**Review by MocanaAtBroad — APPROVED** [review]

**Pull request merged**